WebSocket Origin Restriction (#9447)
aspnet/Docs: aspnetcore/fundamentals/websockets.md

::: moniker-end

::: moniker range="< aspnetcore-2.2"

The following settings can be configured:

* `KeepAliveInterval` - How frequently to send "ping" frames to the client to ensure proxies keep the connection open.

> * `ReceiveBufferSize` - The size of the buffer used to receive data. Advanced users may need to change this for performance tuning based on the size of the data.

Reviewer: What are the default values?

> * `ReceiveBufferSize` - The size of the buffer used to receive data. Advanced users may need to change this for performance tuning based on the size of the data. The default is 4kb.

Reviewer 2: - kb --> KB
- Add space between 4 and KB

::: moniker-end

::: moniker range=">= aspnetcore-2.2"

The following settings can be configured:

* `KeepAliveInterval` - How frequently to send "ping" frames to the client to ensure proxies keep the connection open.
* `ReceiveBufferSize` - The size of the buffer used to receive data. Advanced users may need to change this for performance tuning based on the size of the data.

> * `AllowedOrigins`    - List of the Origin header values allowed for WebSocket requests to prevent Cross-Site WebSocket Hijacking. By default all Origins are allowed.

Reviewer 3: Lowercase Cross-Site and Hijacking

> * `ReceiveBufferSize` - The size of the buffer used to receive data. Advanced users may need to change this for performance tuning based on the size of the data. The default is 4kb.

Reviewer 2: - kb --> KB
- Add space between 4 and KB

> * `AllowedOrigins`    - List of the Origin header values allowed for WebSocket requests to prevent cross-site WebSocket hijacking. By default all Origins are allowed.

Reviewer: Nit: Uppercase O in "Origin header" seems right, but it should probably be lowercase for "By default, all origins are allowed.

> * `AllowedOrigins`    - List of the Origin header values allowed for WebSocket requests to prevent cross-site WebSocket hijacking. By default all Origins are allowed.

Reviewer: ```suggestion
* `AllowedOrigins`    - List of the Origin header values allowed for WebSocket requests to prevent cross-site WebSocket hijacking. By default, all origins are allowed.
```

> * `AllowedOrigins` - List of the Origin header values allowed for WebSocket requests to prevent cross-site WebSocket hijacking. By default, all origins are allowed.

Reviewer 4: ```suggestion
* `AllowedOrigins` - A list of allowed Origin header values for WebSocket requests. By default, all origins are allowed. See "WebSocket origin restriction" below for details
```

It feels like it's best to send people over to the detail description here.

> * `AllowedOrigins`    - List of the Origin header values allowed for WebSocket requests to prevent cross-site WebSocket hijacking. By default, all origins are allowed.

Reviewer 2: ```suggestion
* `AllowedOrigins` - List of the Origin header values allowed for WebSocket requests to prevent cross-site WebSocket hijacking. By default, all origins are allowed.
```

When accepting the WebSocket connection before beginning the loop, the middleware pipeline ends. Upon closing the socket, the pipeline unwinds. That is, the request stops moving forward in the pipeline when the WebSocket is accepted. When the loop is finished and the socket is closed, the request proceeds back up the pipeline.

::: moniker range=">= aspnetcore-2.2"

### WebSocket origin restriction

The protections provided by CORS don't apply to WebSockets. Browsers do **not**:

* Perform CORS pre-flight requests.
* Respect the restrictions specified in `Access-Control` headers when making WebSocket requests.

However, browsers do send the `Origin` header when issuing WebSocket requests. Applications should be configured to validate these headers to ensure that only WebSockets coming from the expected origins are allowed.

> If you are hosting your server on "https://server.com" and hosting your client on "https://client.com", add "https://client.com" to the `AllowedOrigins` list for WebSockets to verify.

Reviewer 2: you are --> you're

> ```c#

Reviewer 2: ```suggestion
```csharp
```

When accepting the WebSocket connection before beginning the loop, the middleware pipeline ends. Upon closing the socket, the pipeline unwinds. That is, the request stops moving forward in the pipeline when the WebSocket is accepted. When the loop is finished and the socket is closed, the request proceeds back up the pipeline.

::: moniker range=">= aspnetcore-2.2"

> ### WebSocket Origin Restriction

Reviewer 3: Lowercase origin restriction

sentence case headings; lowercase "origin" except when it refers specifically to "Origin" as the name of a header.

The protections provided by CORS don't apply to WebSockets. Browsers do **not**:

* Perform CORS pre-flight requests.
* Respect the restrictions specified in `Access-Control` headers when making WebSocket requests.

However, browsers do send the `Origin` header when issuing WebSocket requests. Applications should be configured to validate these headers to ensure that only WebSockets coming from the expected origins are allowed.

If you are hosting your server on "https://server.com" and hosting your client on "https://client.com", add "https://client.com" to the `AllowedOrigins` list for WebSockets to verify.

```c#
app.UseWebSockets(new WebSocketOptions()
{

>     AllowedOrigins.Add("https://client.com");

Reviewer 3: what's the delimiter for multiple origins?

Author: It's a list so you can call `.Add()` for as many origins as you want

Reviewer 4: Show multiple origins (like `https://client.com` and `https://www.client.com`) to illustrate that

});
```

> This helps ensure that browser users are using a trusted client and not a malicious one from a different site.

Author: Review: Should we keep this or remove it?

Reviewer 3: Remove from here, but consider working it into the section intro.

> [!NOTE]

> > The `Origin` header is controlled by the client and, like the `Referer` header, can be faked. These headers should **not** be used as an authentication mechanism.

Reviewer 3: Or more directly

> Do **not** use these headers as an authentication mechanism.

Could also make that sentence first to make it stand out more.
